src/Dictionaries/RegExpTreeDictionary.cpp

```diff
@@ -137,13 +145,60 @@ std::vector<StringPiece> createStringPieces(const String & value, int num_captur
 
 void RegExpTreeDictionary::calculateBytesAllocated()
 {
-    for (const String & regex : regexps)
+    for (const String & regex : simple_regexps)
         bytes_allocated += regex.size();
     bytes_allocated += sizeof(UInt64) * regexp_ids.size();
     bytes_allocated += (sizeof(RegexTreeNode) + sizeof(UInt64)) * regex_nodes.size();
     bytes_allocated += 2 * sizeof(UInt64) * topology_order.size();
 }
 
+namespace
+{
+    /// hyper scan is not good at processing regex containing {0, 200}
+    /// This will make re compilation slow and failed. So we select this heavy regular expressions and
+    /// process it with re2.
+    struct ComplexRegexChecker
+    {
+        re2_st::RE2 searcher;
+        ComplexRegexChecker() : searcher(R"(\{([\d]+),([\d]+)\})") {}
+
+        static bool isFigureLargerThanFifty(const String & str)
+        try
+        {
+            auto number = std::stoi(str);
+            return number > 50;
+        }
+        catch (std::exception &)
+        {
+            return false;
+        }
+
+        [[maybe_unused]]
+        bool check(const String & data) const
```

Comment:
(cosmetic: suggest to rename to `isSimpleRegex()` (+ invert the result). This will make the call prettier, i.e. `if (use_vectorscan && checker.isSimpleRegex(regex))` instead of `if (use_vectorscan && !checker.check(regex))`

Reply:
cosmetic: `bool check(const String & regexp) const`

Reply:
I thought about the re2 vs. hyperscan distinction again. The thing is that hyperscan
1. does not support all of re2's syntax (see [here](https://intel.github.io/hyperscan/dev-reference/compilation.html#pattern-support)),
2. has different matching semantics than re2 (see [here](https://intel.github.io/hyperscan/dev-reference/compilation.html#semantics)) and
3. has different performance characteristics than re2 (ideally faster, sometimes slower)

Class `ComplexRegexChecker` tries to avoid 3. by running a regex over a regex. But this is unfortunately only the tip of the iceberg. My fear is that will be regexes which match differently on re2 vs. hyperscan (case 2.) vs. don't even compile in hyperscan (case 1.). Building comprehensive check that covers all cases is infeasible and achieving good enough test coverage for hyperscan is also quite difficult.

So while the current approach is clever, I am scared that it will at one point in future break.

The [regex matching functions](https://clickhouse.com/docs/en/sql-reference/functions/string-search-functions) in ClickHouse use either re2 OR hyperscan (which is documented in each case) but not both at a time and in particular there are no fallbacks implemented for above reasons.

My suggestion would be the following: RegexpTreeDictionary works by default with re2 syntax. If ClickHouse is compiled with "-DENABLE_HYPERSCAN" (which works btw. only on x86 and ARM) and if setting "regexp_dict_allow_hyperscan" is true, then the RegexpTreeDictionary will evaluate regexes exclusively with hyperscan. There will be no fallback to re2 and in particular no checking if the pattern is fast or slow in HyperScan (--> ComplexRegexChecker). It will be the user's responsibility to provide a Hyperscan-compatible pattern (which can also be evaluated quickly). At the same time, docs of setting "regexp_dict_allow_hyperscan" will note that this option shall be used only at the user's risk / that the setting is experimental.

This will then be in line with the overall way how ClickHouse treats hyperscan right now.

Reply:
Your concern makes sense. However, RegExpTree Dictionary is mainly for UA Parser, which does not likely process very special regex. The difference with hyperscan and re2 seems not easy to cause inconsistent problems. 

If we don't add this fallback thing, almost all of the ua parsers are not able to work under hyperscan :( Let's do it like this util we find real problems